executor/join.go

```diff
@@ -286,34 +304,104 @@ func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.C
 	}
 }
 
-// evalRadixBitNum evaluates the radix bit numbers.
-func (e *HashJoinExec) evalRadixBitNum() {
-	sv := e.ctx.GetSessionVars()
-	// Calculate the bit number needed when using radix partition.
-	if !sv.EnableRadixJoin {
-		return
+// partitionInnerRows re-order e.innerResults into sub-relations.
+func (e *HashJoinExec) partitionInnerRows() error {
+	if err := e.preAlloc4InnerParts(); err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	for i := 0; i < e.partConcurrency; i++ {
+		wg.Add(1)
+		workerID := i
+		go util.WithRecovery(func() {
+			e.doInnerPartition(workerID, wg)
+		}, nil)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) doInnerPartition(startIdx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	chkIdx, chkNum := startIdx, e.innerResult.NumChunks()
+	for ; chkIdx < chkNum; chkIdx += e.partConcurrency {
+		chk := e.innerResult.GetChunk(chkIdx)
+		for rowIdx, partPtr := range e.innerRowPrts[chkIdx] {
+			partIdx, rowPtr := partPtr.partitionIdx, partPtr.RowPtr
+			part := e.innerParts[partIdx]
+			part.Insert(rowPtr, chk.GetRow(rowIdx))
```

Comment:
It just came to my mind that do we really need to cluster the `innerResult` into partitions physically? I mean, our purpose is to make the partitioned hash table fit into cache, so during partition phase, can we just record the `RowPtr`s for a partition, and delay `copy` to the hash table building phase?

Reply:
Our purpose is to reduce the cache miss happened during building and probing hash table. So we did the physical partition.

But I think we can bench the time consumption of the method you mentioned.

Reply:
> can we just record the RowPtrs for a partition?

Probably not. After the inner row pointer is probed from the hash table, we have to join the inner and outer rows to return the result to parent operator. Cache miss may happen if the inner row is not in the L2 Cache. That is to say, the cache miss ratio in the random hash table access is reduced, but the cache miss ratio in the random inner table access still remains without the true partition and copy.